### ISSUE TITLE
specialized support for arrays

### ISSUE THREAD

**Issue by bohrium** [benrbray/prosemirror-math#15]
One of the most annoying-times-frequency things among my latex experiences is making 2d grids by describing them using 1d text.  Such grids include matrices, commutative diagrams, tables, multi-line derivations whose terms are aligned, and more.  It'd be neat to provide a special "array mode" for parsing and editing text of the form 
```
A & B & C & D & E \\  
A & B & C & D & E \\  
A & B & C & D & E \\  
A & B & C & D & E
```
This is the syntax for the latex body in each of all four examples above.  The cells(' latex expressions) will typically all differ in length, creating a headache (do we spend time inserting white-space to align the columns, maintaining this alignment when we modify the cells?  or do we ignore alignment issues in the source code and just count in our head when editing each row?)       

It would be nice if one could edit blocks of the above form in a specialized editor that automatically aligned the ampersands and line-break symbols.

**Comment by benrbray** [comment]
I agree, this is a common pain point for me as well, and I've been mulling over solutions in my head for a while.  Getting the workflow right is tricky, but I'm happy to discuss possibilities.  Although the examples you listed share a common syntax, I think they appear in quite different contexts and will need to be handled differently.  We will also be limited by the commands supported by KaTeX / MathJax.

* Matrices tend to appear as subexpressions of a larger math block, and more than one matrix may appear in a single expression.  How should the editor behave for something like this?

```
A x = \begin{bmatrix}
   a_{11} & a_{12} \\
   a_{21} & a_{22}
\end{bmatrix}
\begin{bmatrix}
	x_1 \\ x_2
\end{bmatrix}
```

* Tables tend to appear by themselves, so it might be as simple as enabling [prosemirror-tables](https://github.com/ProseMirror/prosemirror-tables) extension with some special code for serialization to LaTeX.

* I suspect that everyone has different preferences for how to format multi-line derivations, so it might be difficult to standardize, especially when nested brackets are involved.  Personally I tend to use multiple indented lines for nested brackets, and I'll often wrap manually if lines get too long.  For example,

```
\begin{aligned}
\EE_{q(x)}\left[  (x-\mu_A)^T \Sigma_A^{-1} (x-\mu_A) \right]
&= \EE\left[
        \trace\left( (x-\mu_A)^T \Sigma_A^{-1} (x-\mu_A) \right)
    \right] 
    & \text{(trace of scalar)} \\
&= \EE\left[
        \trace\left( \Sigma_A^{-1} (x-\mu_A) (x-\mu_A)^T \right)
    \right]
    & \text{(cyclicity)}\\
&= \trace\left( \EE\left[
        \Sigma_A^{-1} (x-\mu_A) (x-\mu_A)^T
    \right] \right)
    &\text{(linearity)}\\
&= \trace\left( \Sigma_A^{-1} \EE\left[
        (x-\mu_A) (x-\mu_A)^T
    \right] \right) 
    &\text{(linearity)}\\
&= \trace\left( \Sigma_A^{-1} \left[
        \Sigma_B + (\mu_B - \mu_A)(\mu_B - \mu_A)^T
    \right] \right)
    &\text{(eqn\;\ref{eqn:gauss-quadratic-form})} \\
&= \trace\left( \Sigma_A^{-1} \Sigma_B \right)
    + (\mu_B - \mu_A)^T \Sigma_A^{-1} (\mu_B - \mu_A)
    &\text{(cyclicity; linearity)}
\end{aligned}
```

* Commutative diagrams are tricky, because neither KaTeX nor MathJax has great support for them out of the box.  It would be helpful to collect some examples of source code for commutative diagrams that we expect `prosemirror-math` to be able to handle.  The best tool I know of for editing commutative diagrams on the web is [quiver](https://q.uiver.app/), so worst case we might be able to introduce a new type of math node that makes it easy to embed diagrams made with quiver.

**Comment by bohrium** [comment]
Cool!  Your points about differences in context implying differences in handling make a lot of sense to me.  I'll mull over.  

Quick comments: quiver seems to generate tikz blocks, but katex does not support the tikz environments (I think).  It seems they support the amscd "CD" environment (CD stands for commutative diagram), which is a very bare bones package that can express perhaps 80% of the simple diagrams a math student might draw and that can render them correctly perhaps 80% of the time.

**Comment by bohrium** [comment]
As for a basic set of diagrams that would be nice to handle, here are what come to mind:

**lifting diagram**
```
\begin{tikzcd}
                               & P \arrow[d, "g"] \arrow[ld, "l"', dotted] \\
X \arrow[r, "\pi"', two heads] & Y                                        
\end{tikzcd}
```
[see this link](https://tikzcd.yichuanshen.de/#N4Igdg9gJgpgziAXAbVABwnAlgFyxMJZABgBoBGAXVJADcBDAGwFcYkQANEAX1PU1z5CKcqWLU6TVuwAKPPiAzY8BIqKo0GLNohABNHhJhQA5vCKgAZgCcIAWySiQOCEgBMmqTpAn5V2w6ITi5IZCCMWGDeUBA4OMYgntrsjInh9ABGMIwyAirCINZYJgAWOH4gNvahNCGIHiAlMPRQ7DgA7hBNLQhJ0roAOgNoWGmMmdm5ykLsRaXl3JTcQA)

**causal model**
```
\begin{tikzcd}
\substack{\text{sensor}\\ \text{placement}} \arrow[rd] \arrow[rrd] &                                                     &                                              \\
                                                                   & \substack{\text{sensor}\\ \text{reading}} \arrow[r] & \substack{\text{guessed}\\ \text{heartrate}} \\
\substack{\text{true}\\ \text{heartrate}} \arrow[ru]               &                                                     &                                             
\end{tikzcd}
```
[see this link](https://tikzcd.yichuanshen.de/#N4Igdg9gJgpgziAXAbVABwnAlgFyxMJZARgBpiBdUkANwEMAbAVxiRAB124mAjOHOgGMA1sE44YADxzA4MMHAgAnAL6dOAAnFSZSmHShYwAcxUqQK0uky58hFACZyVWoxZtO3PgJFj2E6WBjFjg5KDV1LX8dYAALfSUcJToJMwsrEAxsPAIiAAZSPJd6ZlZEDi5efiFRbUC5BWUIzTqZNAYhGABbeRw0y2tsu3zSB2K3Moqvat9W4CSWZvYogJl4ukTk1PMVFxgoY3giUAAzJQgupAKQHAgkYgGQM4ukJxu7xDzH58vEAGZqLcrt9zr83kDEA8KCogA)

**Comment by bohrium** [comment]
Let me know if any of these ideas sounds like a good starting point that I might try to implement!

**Comment by benrbray** [comment]
I think we agree on the rough ux, so I'm happy to continue the feedback loop with you if you start on a prototype!  Some things to ponder:

* I plan to add support for MathJax, for those who want it, along with extensions like [XyJax](https://sonoisa.github.io/xyjax/xyjax.html) for commutative diagrams.  It's unfortunate that the KaTeX `\begin{CD}` environment doesn't support diagonal arrows.  So, you can design with something like XyJax in mind.

* How do you plan to implement this functionality?  As an actual `<table>` element that gets converted to LaTeX?  Or as a string processor that adjusts the contents of the text input field automatically as you type?  We'll need to think carefully about the right way for a user to "activate" the array-style editing mode, since having it on by default might confuse users.  One option is to create a new [`NodeView`](https://github.com/benrbray/prosemirror-math/blob/master/src/math-nodeview.ts#L36) activated by a different character sequence (like `$$matrix` or `\begin{matrix}` etc) that offers the array-style mode.  You can use the [existing math NodeView](https://github.com/benrbray/prosemirror-math/blob/master/src/math-nodeview.ts#L36) as a reference.

* In order to contribute code to this repo, it will be important to understand the transaction-based model of ProseMirror.  The [ProseMirror guide](https://prosemirror.net/docs/guide/) is a good starting point.  Right now, the math editor appears in a [NodeView](https://prosemirror.net/docs/ref/#view.NodeView) that contains another mini-instance of ProseMirror.  So, if you make any changes to how the inner editor behaves, they will need to respect the transactional model.  (meaning you can't just set e.g. `textField.textContent = adjustSpacing(textField.textContent)` every time you want to make a change)
